extensions/sql-database-projects/src/dialogs/addDatabaseReferenceDialog.ts

```diff
@@ -0,0 +1,403 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import * as vscode from 'vscode';
+import * as constants from '../common/constants';
+
+import { Project, SystemDatabase, DatabaseReferenceLocation } from '../models/project';
+import { cssStyles } from '../common/uiConstants';
+import { IconPathHelper } from '../common/iconHelper';
+import { ISystemDatabaseReferenceSettings, IDacpacReferenceSettings } from '../models/IDatabaseReferenceSettings';
+import { isEmptyOrUndefined } from '../common/utils';
+
+export enum ReferenceType {
+	project,
+	systemDb,
+	dacpac
+}
+
+interface Deferred<T> {
+	resolve: (result: T | Promise<T>) => void;
+	reject: (reason: any) => void;
+}
```

Comment:
Can this be done with normal Promises?  I think we should be trying to remove Deferred bits where possible.

Reply:
Normal promise won't work here. The problem isn't really with the source code but with the test code.
The content won't be registered until wizard.open or wizard.start has been called. So, to say this.addDatabaseReferenceTab.registerContent wouldn't finish until azdata.window.openDialog(this.dialog) is called.
So for the tests to know exactly when the content finished registering, we need to have a deferred promise.